packages/flutter_tools/lib/src/dart/pub.dart

```diff
@@ -439,4 +458,27 @@ class _DefaultPub implements Pub {
     }
     return environment;
   }
+
+  /// Insert the flutter_gen synthetic package into the package configuration file if
+  /// there is an l10n.yaml.
+  Future<void> _updatePackageConfig(File packageConfigFile, Directory generatedDirectory) async {
+    if (!packageConfigFile.existsSync()) {
+      return;
+    }
+    final PackageConfig packageConfig = await loadPackageConfigWithLogging(packageConfigFile, logger: _logger);
+    final Package flutterGen = Package('flutter_gen', generatedDirectory.uri, languageVersion: LanguageVersion(2, 8));
+    if (packageConfig.packages.any((Package package) => package.name == 'flutter_gen')) {
+      return;
+    }
+    final PackageConfig newPackageConfig = PackageConfig(
+      <Package>[
+        ...packageConfig.packages,
+        flutterGen,
+      ],
+    );
+    // There is no current API for saving a package_config without hitting the real filesystem.
+    if (packageConfigFile.fileSystem is LocalFileSystem) {
+      await savePackageConfig(newPackageConfig, packageConfigFile.parent.parent);
```

Comment:
`savePackageConfig` is a utility function written in terms of public APIs:

https://github.com/dart-lang/package_config/blob/9c586d04bd26fef01215fd10e7ab96a3050cfa64/lib/src/package_config_io.dart#L131

Maybe the tool should do this more manually to avoid the possibly brittle filesystem type check.

Reply:
Neither writePackageConfigJsonUtf8 nor any of the other utility methods are exported

Reply:
Oh, whoops. I misread. I thought I saw it using https://pub.dev/documentation/package_config/latest/package_config.package_config_discovery/PackageConfig/writeBytes.html

Reply:
You could also locally override dart:io with an IOOverrides, but that is possibly not actually an improvement over the type check here.